File: package/main/src/Error/retry.ts
export interface RetryOptions {
  retries?: number;
  delay?: number;
  shouldRetry?: (error: unknown) => boolean;
}

/**
 * Retries a given async function with configurable retry logic
 *
 * @template T - The return type of the function
 * @param function_ - The async function to retry
 * @param options - Configuration options for retry behavior
 * @param options.retries - Maximum number of retry attempts (default: 3)
 * @param options.delay - Delay between retries in milliseconds (default: 1000)
 * @param options.shouldRetry - Function to determine if an error should trigger a retry (default: always retry)
 * @returns Promise that resolves with the function result or rejects with the final error
 *
 * @example
 * ```typescript
 * // Basic usage with default options
 * const result = await retry(async () => {
 *   const response = await fetch('/api/data');
 *   if (!response.ok) throw new Error('Failed to fetch');
 *   return response.json();
 * });
 *
 * // Custom retry configuration
 * const result = await retry(
 *   () => riskyOperation(),
 *   {
 *     retries: 5,
 *     delay: 2000,
 *     shouldRetry: (error) => error instanceof NetworkError
 *   }
 * );
 * ```
 */
export const retry = <T>(
  function_: () => Promise<T>,
  options: RetryOptions = {},
): Promise<T> => {
  const { retries = 3, delay = 1000, shouldRetry = () => true } = options;
  const attempt = async (remainingAttempts: number): Promise<T> => {
    try {
      return await function_();
    } catch (error) {
      if (remainingAttempts <= 0 || !shouldRetry(error)) {
        throw error;
      }
      return await new Promise((resolve) =>
        setTimeout(() => resolve(attempt(remainingAttempts - 1)), delay),
      );
    }
  };
  return attempt(retries);
};

File: package/umt_python/src/error/retry.py
import asyncio
from typing import Callable, TypeVar, Awaitable, Optional

T = TypeVar("T")


async def retry(
    fn: Callable[[], Awaitable[T]],
    retries: int = 3,
    delay: float = 1.0,
    should_retry: Optional[Callable[[Exception], bool]] = None,
) -> T:
    """
    Retries a given async function with configurable retry logic.

    Args:
        fn: The async function to retry
        retries: Maximum number of retry attempts (default: 3)
        delay: Delay between retries in seconds (default: 1.0)
        should_retry: Function to determine if an error should trigger a retry (default: always retry)

    Returns:
        The result of the function

    Raises:
        The final exception if all retries fail

    Example:
        >>> async def risky_operation():
        ...     return await fetch_data()
        >>> result = await retry(risky_operation, retries=5, delay=2.0)
    """

    def _always_retry(_: Exception) -> bool:
        return True

    if should_retry is None:
        should_retry = _always_retry

    async def attempt(remaining_attempts: int) -> T:
        try:
            return await fn()
        except Exception as error:
            if remaining_attempts <= 0 or not should_retry(error):
                raise
            await asyncio.sleep(delay)
            return await attempt(remaining_attempts - 1)

    return await attempt(retries)
